### PR TITLE
Implement EIP-7549

### PR DESCRIPTION
**Pull request by eserilev**
## Issue Addressed

#5130

### CODE REVIEW

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 1]
```diff
@@ -1934,17 +1934,36 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             };
         drop(cache_timer);
 
-        Ok(Attestation {
-            aggregation_bits: BitList::with_capacity(committee_len)?,
-            data: AttestationData {
-                slot: request_slot,
+        match self.spec.fork_name_at_slot::<T::EthSpec>(request_slot) {
```
Comment (dapplion) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: @michaelsproul are you sure Ord of ForkName won't look cool? :)

File: beacon_node/http_api/tests/tests.rs [hunk 2]
```diff
@@ -3271,6 +3271,7 @@ impl ApiTester {
 
         let mut attestation = Attestation {
             aggregation_bits: BitList::with_capacity(duty.committee_length as usize).unwrap(),
+            index: <_>::default(),
```
Comment (dapplion) [inline beacon_node/http_api/tests/tests.rs]: Setting it to the value `0` feels more explicit than using default. use .into() or the type new()

File: beacon_node/beacon_chain/src/attestation_verification.rs [hunk 3]
```diff
@@ -559,8 +560,21 @@ impl<'a, T: BeaconChainTypes> IndexedAggregatedAttestation<'a, T> {
...
 
-                get_indexed_attestation(committee.committee, attestation)
-                    .map_err(|e| BeaconChainError::from(e).into())
+                match attestation {
+                    Attestation::Base(att) => {
+                        indexed_attestation_base::get_indexed_attestation(committee.committee, att)
+                            .map_err(|e| BeaconChainError::from(e).into())
+                    }
+                    Attestation::Electra(att) => {
+                        // TODO(eip7594) need to get access to the beacon state
```
Comment (dapplion) [inline beacon_node/beacon_chain/src/attestation_verification.rs]: Why access to the full state? You just need the entire shuffling

File: beacon_node/beacon_chain/src/naive_aggregation_pool.rs [hunk 4]
```diff
@@ -125,10 +125,11 @@ impl<E: EthSpec> AggregateMap for AggregatedAttestationMap<E> {
         let _timer = metrics::start_timer(&metrics::ATTESTATION_PROCESSING_AGG_POOL_CORE_INSERT);
 
         let set_bits = a
-            .aggregation_bits
+            .aggregation_bits()
             .iter()
             .enumerate()
-            .filter(|(_i, bit)| *bit)
+            // TODO(eip7594) double check this, it used to be a bit
+            .filter(|(_i, &bit)| bit == 1u8)
```
Comment (dapplion) [inline beacon_node/beacon_chain/src/naive_aggregation_pool.rs]: It is still a bit right? Can you link spec where it switched to u8?

File: consensus/state_processing/src/consensus_context.rs [hunk 5]
```diff
@@ -43,6 +57,7 @@ impl<E: EthSpec> ConsensusContext<E> {
             proposer_index: None,
             current_block_root: None,
             indexed_attestations: HashMap::new(),
+            indexed_attestations_electra: HashMap::new(),
```
Comment (dapplion) [inline consensus/state_processing/src/consensus_context.rs]: Consider coercing everything to electra indexed attestations

File: consensus/state_processing/src/per_block_processing/process_operations.rs [hunk 6]
```diff
@@ -71,13 +72,13 @@ pub mod base {
             .map_err(|e| e.into_with_index(i))?;
 
             let pending_attestation = PendingAttestation {
-                aggregation_bits: attestation.aggregation_bits.clone(),
-                data: attestation.data.clone(),
-                inclusion_delay: state.slot().safe_sub(attestation.data.slot)?.as_u64(),
+                aggregation_bits: attestation.aggregation_bits_base().unwrap().clone(),
```
Comment (dapplion) [inline consensus/state_processing/src/per_block_processing/process_operations.rs]:
```suggestion
                aggregation_bits: attestation.aggregation_bits_base()?.clone(),
```

File: consensus/state_processing/src/per_block_processing/process_operations.rs [hunk 7]
```diff
@@ -129,13 +135,110 @@ pub mod altair_deneb {
...
+            .safe_sub(PROPOSER_WEIGHT)?
+            .safe_mul(WEIGHT_DENOMINATOR)?
+            .safe_div(PROPOSER_WEIGHT)?;
+        let proposer_reward = proposer_reward_numerator.safe_div(proposer_reward_denominator)?;
+        increase_balance(state, proposer_index as usize, proposer_reward)?;
+        Ok(())
+    }
+}
+
+pub mod electra {
```
Comment (dapplion) [inline consensus/state_processing/src/per_block_processing/process_operations.rs]: Not necessary to duplicate all this logic. The only difference is when converting an on-chain attestation to an indexed attestation which can be common in all forks. Just convert attestations from any fork into the electra indexed attestation. The relevant attestation data for reward processing is the same too.

File: consensus/types/src/beacon_block.rs [hunk 8]
```diff
@@ -627,7 +632,8 @@ impl<E: EthSpec, Payload: AbstractExecPayload<E>> BeaconBlockElectra<E, Payload>
             body: BeaconBlockBodyElectra {
                 proposer_slashings: base_block.body.proposer_slashings,
                 attester_slashings: base_block.body.attester_slashings,
-                attestations: base_block.body.attestations,
+                // TODO(eip7549) is setting to default ok here?
+                attestations: <_>::default(),
```
Comment (dapplion) [inline consensus/types/src/beacon_block.rs]: No, that is not representative of the maximum size. You need to create a list with max count of electra attestations of maximum size

File: consensus/types/src/indexed_attestation.rs [hunk 9]
```diff
@@ -12,25 +15,51 @@ use tree_hash_derive::TreeHash;
...
 #[arbitrary(bound = "E: EthSpec")]
 pub struct IndexedAttestation<E: EthSpec> {
     /// Lists validator registry indices, not committee indices.
+    #[superstruct(only(Base), partial_getter(rename = "attesting_indices_base"))]
     #[serde(with = "quoted_variable_list_u64")]
     pub attesting_indices: VariableList<u64, E::MaxValidatorsPerCommittee>,
+
+    #[superstruct(only(Electra), partial_getter(rename = "attesting_indices_electra"))]
+    #[serde(with = "quoted_variable_list_u64")]
+    pub attesting_indices: VariableList<u64, E::MaxValidatorsPerCommitteePerSlot>,
```
Comment (dapplion) [inline consensus/types/src/indexed_attestation.rs:33]: oooohhh that's annoying.. Hmm in practice this change does not affect much as it strictly increases the maximum length of the list. This is only relevant when merkleizing attester slashings. In LH internals we can have a common type for both or just use the bigger electra variant everywhere, and downsize to the base when doing block inclusions

File: beacon_node/beacon_chain/src/attestation_verification.rs (hunk 3 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/attestation_verification.rs]: I chatted about this with Eitan on Tuesday and recommended plumbing through the committee cache with the full set of committees. Just requires a bit of tweaking of `map_committee_cache`

File: consensus/types/src/indexed_attestation.rs (hunk 9 above)
Comment (eserilev) [reply consensus/types/src/indexed_attestation.rs:33]: im now using the electra variant everywhere, though i still need to add logic to downsize & enfore the smaller list size for the base variant

File: beacon_node/beacon_chain/src/naive_aggregation_pool.rs (hunk 4 above)
Comment (eserilev) [reply beacon_node/beacon_chain/src/naive_aggregation_pool.rs]: ive reverted this snippet of code back to its original

File: consensus/state_processing/src/per_block_processing/process_operations.rs [hunk 10]
```diff
@@ -73,14 +76,25 @@ pub mod base {
...
-                proposer_index,
+            let pending_attestation = match attestation {
+                Attestation::Base(att) => PendingAttestation::Base(PendingAttestationBase {
+                    aggregation_bits: att.aggregation_bits.clone(),
+                    data: att.data.clone(),
+                    inclusion_delay: state.slot().safe_sub(att.data.slot)?.as_u64(),
+                    proposer_index,
+                }),
+                Attestation::Electra(att) => {
+                    PendingAttestation::Electra(PendingAttestationElectra {
```
Comment (michaelsproul) [inline consensus/state_processing/src/per_block_processing/process_operations.rs:87]: this should just an error

File: consensus/types/src/pending_attestation.rs [hunk 11]
```diff
@@ -1,34 +1,124 @@
...
+    Deserialize,
     arbitrary::Arbitrary,
+    PartialEq,
 )]
+#[serde(untagged)]
+#[tree_hash(enum_behaviour = "transparent")]
+#[ssz(enum_behaviour = "transparent")]
+#[serde(bound = "E: EthSpec", deny_unknown_fields)]
 #[arbitrary(bound = "E: EthSpec")]
 pub struct PendingAttestation<E: EthSpec> {
```
Comment (michaelsproul) [inline consensus/types/src/pending_attestation.rs:53]: this doesn't need be superstructed (it is phase0 only)

**Pull request closed without merging**